mlir/lib/Dialect/EmitC/Transforms/EliminateLibm.cpp

```diff
@@ -0,0 +1,122 @@
+// prototypes (e.g., inserted by the MathToLibm pass) by an inclusion of the
+// <math.h> header.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/EmitC/IR/EmitC.h"
+#include "mlir/Dialect/EmitC/Transforms/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/Visitors.h"
+#include "mlir/Interfaces/FunctionInterfaces.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace emitc {
+#define GEN_PASS_DEF_ELIMINATELIBM
+#include "mlir/Dialect/EmitC/Transforms/Passes.h.inc"
+} // namespace emitc
+} // namespace mlir
+
+using namespace mlir;
+using namespace emitc;
+
+namespace {
+
+/// Replace all Libm calls (where callee has `libm` attribute + no definition)
+/// by opaque calls
+struct OpacifyLibmCall : public OpRewritePattern<func::CallOp> {
+  using OpRewritePattern<func::CallOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(func::CallOp callOp,
+                                PatternRewriter &rewriter) const override {
+
+    auto *st = SymbolTable::getNearestSymbolTable(callOp);
+    auto *calleeSym = SymbolTable::lookupSymbolIn(st, callOp.getCallee());
+    assert(isa<FunctionOpInterface>(calleeSym));
+    FunctionOpInterface callee = dyn_cast<FunctionOpInterface>(calleeSym);
+
+    if (!(callee->hasAttr("libm") && callee.isDeclaration()))
+      return failure();
+
+    auto opaqueCall = rewriter.create<emitc::CallOpaqueOp>(
+        callOp->getLoc(), callOp.getResultTypes(), callOp.getCallee(),
+        callOp.getArgOperands());
+
+    rewriter.replaceOp(callOp, opaqueCall);
+    return success();
+  }
+};
+
+struct EliminateLibmPass
+    : public emitc::impl::EliminateLibmBase<EliminateLibmPass> {
+  void runOnOperation() override {
+    Operation *rootOp = getOperation();
+    MLIRContext *context = rootOp->getContext();
+
+    if (!llvm::isa<ModuleOp>(rootOp))
+      return;
+
+    ModuleOp module = dyn_cast<ModuleOp>(rootOp);
+
+    // Find the first math.h inclusion
+    SmallVector<func::FuncOp> libmPrototypes;
+    module.walk([&libmPrototypes](func::FuncOp funcOp) {
+      if (funcOp->hasAttr("libm") && funcOp.isDeclaration())
+        libmPrototypes.push_back(funcOp);
+    });
+
+    if (libmPrototypes.empty())
+      return;
+
+    // Replace with a rewrite pattern
+    RewritePatternSet opacifyLibmCallPatterns(context);
+    opacifyLibmCallPatterns.add<OpacifyLibmCall>(
+        opacifyLibmCallPatterns.getContext());
+
+    if (failed(applyPatternsAndFoldGreedily(
+            rootOp, std::move(opacifyLibmCallPatterns))))
+      return signalPassFailure();
+
+    // Check that none of the prototypes have users
+    for (auto &proto : libmPrototypes) {
+      assert(proto->getUsers().empty());
```

Comment:
I don't think the pass guarantees this property, so I suggest to instead `return signalPassFailure();`